src/main.cpp

```diff
@@ -53,27 +46,32 @@ int main(int argc, char** argv)
     }
 
     std::cout << "Waiting to discover system..." << std::endl;
-    dc.register_on_discover([&discovered_system](uint64_t uuid) {
-        std::cout << "Discovered system with UUID: " << uuid << std::endl;
-        discovered_system = true;
-    });
+    auto prom = std::make_shared<std::promise<void>>();
+    std::future<void> fut = prom->get_future();
 
-    // We usually receive heartbeats at 1Hz, therefore we should find a system after around 2
-    // seconds.
-    sleep_for(seconds(2));
+    System& system = dc.system();
+    system.register_component_discovered_callback([prom, &system](ComponentType component_type) {
+        std::cout << NORMAL_CONSOLE_TEXT << "Discovered a component with type "
+                  << unsigned(component_type) << std::endl;
+        try {
+            prom->set_value();
+            // We only need to receive this once, now we can unsubscribe again.
+            system.register_component_discovered_callback(nullptr);
+        } catch (const std::future_error& e) {
+            // Ignore if prom is set multiple times which can happen even though we're
```

Comment:
Is this still an issue if you unregister before triggering the promise?

Reply:
Unfortunately yes if the callback gets called in rapid succession.